### ISSUE TITLE
[Questions] 關於Model Head 以及 LLM Training dataset的請教

### ISSUE THREAD

**Issue by smithlai** [zyds/transformers-code#13]
您好，我想問兩個問題
第一個是關於 "AutoModelForCausalLM.from_pretrained"     
我知道這個在基礎課程with/without Head篇章有提到過，基本上使用AutoModel會提取底層的Bert，輸出則為Hidden State。    
使用ForCausalLM則會在後續加上Fully Connected Layer之類的，用以生成logits完成下游任務。     
在Bert我可以理解，因為他本來就是將文字embedding成hidden state的過程，沒有下游任務。但是以bloom而言，     
AutoModelForCausalLM也會在他後續又加上Header嗎？我就是在這邊一直糾結。     
或是反過來，對於本來就有下游任務的某個model，比如llama2-instruct，那我用AutoModelForCausalLM跟AutoModel去讀取他，會有什麼差別？    
AutoModel是會讀取完整模型，還是會把他的Head給切了？還是AutoModelForCausalLM會又額外給他多加新的head layer?    

第二個問題是關於 Training Data是使用causal data(['input_ids']+DataCollatorForCausalLM)比較好，還是seq2seq data(['input_ids']+['label']+DataCollatorFromSeq2Seq)？    
是取決於什麼呢？取決於模型嗎？    
比如gpt-2就應該使用causal data,  T5就應該使用seq2seq data?    
    
還是取決於任務跟效果？     
比如是希望模型能回答出一模一樣的結果使用seq2seq?    

我會有這個疑問是因為看到其他人的程式碼 [04-Gemma-2-9b-it peft lora微调.ipynb](https://github.com/datawhalechina/self-llm/blob/master/models/Gemma2/04-Gemma-2-9b-it%20peft%20lora%E5%BE%AE%E8%B0%83.ipynb)  以及 [fine_tune_gemma_model.ipynb](https://github.com/NSTiwari/Gemma-on-Android/blob/main/fine_tune_gemma_model.ipynb)      
明明同樣是Gemma，其中一個卻使用seq2seq讓我感到困惑

**Comment by martain3** [comment]
https://percent4.github.io/NLP%EF%BC%88%E4%B9%9D%E5%8D%81%E5%9B%9B%EF%BC%89transformers%E6%A8%A1%E5%9D%97%E4%B8%AD%E7%9A%84DataCollator/  希望这个帖子能帮助理解第二个问题。

**Comment by martain3** [comment]
对于问题一，AutoModel会加载模型的基础部分，即它只会加载模型的主干（backbone），不包括任何任务特定的头层（head layer），我认为是砍掉了。这仅仅是我个人看法。具体最好还是看源码。

**Comment by smithlai** [comment]
第二個問題我可能要化點時間消化，先說聲感謝。

第一個問題我剛好這兩天花了幾天看程式碼加上ChatGPT，剛剛才看出一點心得。

首先每個Model都是某基本型態(比如Bert 或 T5)，ForXXXX就是又疊加Layer(Head)上去。

這些Model with Head基本上形狀是永遠固定的，比如ForSeqClassifiaction就是基本型疊上一層classifier

這些層並不是由pretrained複製過來的，更正確來說，是我們建立好整個模型的層，再將權重複製進來。

我之前的誤解在於，我一直以為Model或是Layer是經由from_pretrain複製過來，但事實上Model在Class實體化時形狀就固定了，被from_pretrain複製的僅有權重。

如果用AutoModel，就會只建立基本型態Backbone(如BERT)，from_pretrained再將權重複製過來
如果用AutoModelForSeqClassifiaction，就會建立初始化基本型疊上一層classifier，from_pretrained再將權重複製過來


也就是說：
AutoModel即使讀取了有下游任務的模型，自身也只有基本型態可以被賦值，等同砍掉Head
AutoModelForSeqClassifiaction讀取了基本型態如Bert，也只會複製到基本型態的權重，head沒能得到權重需要訓練。
如果讀取的是相同的型態，則權重完全複製。
如果是不相符的下游任務，比如AutoModelForSeqClassifiaction 但 num_labels不同，則會出現shape mismatch的exception

**Comment by smithlai** [comment]
關於第二個問題，我閱讀過之後知道了DataCollator是針對不同的需求做了Padding或Labeling。而ForSeq2seq跟ForLanguageModeling 基本相同，主要的差別是seq2seq對於前半Label會使用-100, 而後半使用input_id，而ForLanguageModeling則是通通填充input_id

我的疑問是這樣的：Training同一種model，他們卻用了兩種不同的資料格式，所以讓我感到困惑。


[第一組選手](https://github.com/smithlai/Gemma-on-Android/blob/main/fine_tune_gemma_model.ipynb)使用最基本的tokenizer(sen1,sen2)：    

`data = data.map(lambda samples: tokenizer(samples["Question"], samples["Context"]), batched=True)`     

[第二組選手](https://github.com/datawhalechina/self-llm/blob/master/models/Gemma2/04-Gemma-2-9b-it%20peft%20lora%E5%BE%AE%E8%B0%83.ipynb)使用手動加label + DataCollatorForSeq2seq：  

```py
def process_func(example):
    MAX_LENGTH = 384    # Llama分词器会将一个中文字切分为多个token，因此需要放开一些最大长度，保证数据的完整性
    input_ids, attention_mask, labels = [], [], []
    instruction = tokenizer(f"<bos><start_of_turn>user\n{example['instruction'] + example['input']}<end_of_turn>\n<start_of_turn>model\n", add_special_tokens=False)  # add_special_tokens 不在开头加 special_tokens
    response = tokenizer(f"{example['output']}<end_of_turn>\n", add_special_tokens=False)
    input_ids = instruction["input_ids"] + response["input_ids"] + [tokenizer.pad_token_id]
    attention_mask = instruction["attention_mask"] + response["attention_mask"] + [1]  # 因为eos token咱们也是要关注的所以 补充为1
    labels = [-100] * len(instruction["input_ids"]) + response["input_ids"] + [tokenizer.pad_token_id]  
    if len(input_ids) > MAX_LENGTH:  # 做一个截断
        input_ids = input_ids[:MAX_LENGTH]
        attention_mask = attention_mask[:MAX_LENGTH]
        labels = labels[:MAX_LENGTH]
    return {
        "input_ids": input_ids,
        "attention_mask": attention_mask,
        "labels": labels
    }
```
 
第三組(略)： 使用DataCollatorForLanguageModeling

------------------


第一個疑問：
這幾種訓練方式對於訓練本身會有所不同嗎？
如果是相同的，那意思是第1組會在底層再偷偷把label加上去嗎？
抑或相反，是Label在第2.3組根本沒有作用?

第二個疑問：
如果label是必須的，那為什麼要分ForSeq2seq跟ForLanguageModeling？我們既然想Finetune，那當然是訓練回答，感覺用Seq2Seq著重後半句如何回答的才合理啊？
我就是在這邊感覺一般的CasualLM跟seq2seq的界線又模糊了，所以延伸第三個問題

第三個疑問：
對於非seq2seq的Model(如gemma)，為什麼他還能塞seq2seq去訓練。那是不是反過來我也可以用ForLanguageModeling去訓練seq2seq model?

很抱歉問得很混亂，因為我這邊關於model的運作還是似懂非懂的，尤其是他們對於訓練資料的使用。
求大佬指導。

**Comment by zyds** [comment]
@smithlai 

第一个疑问：第一组会加上去，如果你可以看一下SFTTrainer的源码，可以看到他默认会使用DataCollatorForLanguageModeling，通过这个把labels补回来，这个是必须的

第二个疑问：正如你表述的，我们主要还是为了训练回复部分，ForLanguageModeling是为了原生的语言模型的训练，边界也不算模糊，只是数据填充的格式

第三个疑问：核心在于，你到底要训练什么，不要被DataCollator的命名带着走，事实上完全也可以不使用预先提供的这些DataCollator，按照你自己的想法进行label的设置也是可以的，只要是能正常训练，回复中照样也可以加-100，就看你的目标啦，但是到底是哪种方式训练出来的更好，这个就得看实验了

这部分可以回头再看一下，实战篇各个章节使用的DataCollator，他们到底将数据填充至了什么格式，以及为什么要这样填充

**Comment by smithlai** [comment]
@zyds 感謝大佬條理分明的指點，我好像悟了。
